### PR TITLE
NCSD-848: Rewrite rules & Custom probe

### PR DESCRIPTION
**Pull request by rivoric**
Added
* Optional custom probe config
* Optional rewrite rules

### CODE REVIEW

File: Resources/ArmTemplates/app-gateway.json [hunk 1]
```diff
@@ -154,8 +175,8 @@
                 }
             }
         },
-        "useSslCerts": "[and(greater(length(parameters('keyVaultName')),0), greater(length(parameters('keyVaultSecretName')),0), greater(length(parameters('userAssignedIdentityName')),0))]",
-        "noSslCerts": [],
+        "useSslCerts": "[and(not(empty(parameters('keyVaultName'))), not(empty(parameters('keyVaultSecretName'))), not(empty(parameters('userAssignedIdentityName'))))]",
```
Comment (NickGraham101) [inline Resources/ArmTemplates/app-gateway.json]: I thought we'd decided that greater than length 0 was a better way to evaluate this (after I wrote some horribly convoluted statement)

File: Resources/ArmTemplates/app-gateway.md [hunk 2]
```diff
@@ -20,7 +20,17 @@ appGatewayTier: (optional) string
...
 Must be one of Standard_Small (default if none supplied), Standard_Medium, Standard_Large, WAF_Medium, WAF_Large, Standard_v2 or WAF_v2
-                
+
+useCustomProbe: (optional) boolean
+
+Use a custom probe. Defaults to false (no custom probe).
+
+customProbeProperties: (optional) object
+
+Override default probe properties or set additional custom probe properties.
```
Comment (NickGraham101) [inline Resources/ArmTemplates/app-gateway.md]: Has this been tested?  The documentation on the union function is ambiguous about how it works in this situation, it isn't clear if the first or the 2nd property is kept

File: Resources/ArmTemplates/app-gateway.json [hunk 3]
```diff
@@ -286,7 +338,7 @@
                 "frontendIPConfigurations": [
                     {
                         "name": "appGatewayFrontendIP",
-                        "properties": "[if(and(greater(length(parameters('privateIpAddress')),0),greater(length(parameters('publicIpAddressId')),0)), union(variables('frontendIp').public,variables('frontendIp').private), if(greater(length(parameters('privateIpAddress')),0), variables('frontendIp').private, if(greater(length(parameters('publicIpAddressId')),0), variables('frontendIp').public, json('null'))))]"
+                        "properties": "[if(and(not(empty(parameters('privateIpAddress'))), not(empty(parameters('publicIpAddressId')))), union(variables('frontendIp').public,variables('frontendIp').private), if(not(empty(parameters('privateIpAddress'))), variables('frontendIp').private, if(not(empty(parameters('publicIpAddressId'))), variables('frontendIp').public, json('null'))))]"
```
Comment (NickGraham101) [inline Resources/ArmTemplates/app-gateway.json]: I thought we'd decided that greater than length 0 was a better way to evaluate this? (after I wrote some horribly convoluted statement)

File: Resources/ArmTemplates/app-gateway.md (hunk 2 above)
Comment (rivoric) [reply Resources/ArmTemplates/app-gateway.md]: The example in the documentation says it second overwrites the first - https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-array#union

File: Resources/ArmTemplates/app-gateway.json (hunk 1 above)
Comment (rivoric) [reply Resources/ArmTemplates/app-gateway.json]: Reverted back

File: Resources/ArmTemplates/app-gateway.json (hunk 3 above)
Comment (rivoric) [reply Resources/ArmTemplates/app-gateway.json]: Reverted back

**Pull request merged**